cishouseholds/pipeline/run.py

```diff
@@ -45,25 +48,29 @@ def run_from_config():
     add_run_status(run_id, "finished")
 
 
-def run_pipeline_stages(pipeline_stage_list: list, run_id: int):
+def run_pipeline_stages(pipeline_stage_list: list, run_id: int, retry_count: int, retry_wait_time: int):
     """Run each stage of the pipeline. Catches, prints and logs any errors, but continues the pipeline run."""
     number_of_stages = len(pipeline_stage_list)
     max_digits = len(str(number_of_stages))
     pipeline_error_count = 0
     for n, stage_config in enumerate(pipeline_stage_list):
         stage_start = datetime.now()
-        try:
-            stage_name = stage_config.pop("function")
-            stage_text = f"Stage {n + 1 :0{max_digits}}/{number_of_stages}: {stage_name}"
-            print(stage_text)  # functional
-            pipeline_stages[stage_name](**stage_config)
-        except Exception:
-            pipeline_error_count += 1
-            add_run_status(run_id, "errored", stage_text, "\n".join(traceback.format_exc()))
-            print(f"Error: {traceback.format_exc()}")  # functional
-        finally:
-            run_time = (datetime.now() - stage_start).total_seconds()
-            print(f"    - completed in: {run_time//60:.0f} minute(s) and {run_time%60:.1f} second(s)")  # functional
+        for attempt in range(1, retry_count + 1):
+            try:
+                stage_name = stage_config.pop("function")
+                stage_text = f"Stage {n + 1 :0{max_digits}}/{number_of_stages}: {stage_name}"
+                print(stage_text)  # functional
+                pipeline_stages[stage_name](**stage_config)
+                break
```

Comment:
Have you checked whether this `break` breaks out the try/except instead of the for loop?

An alternative for the loop would be to use `while (attempt_number < retry_count) and not succeeded:`, then setting succeeded to `True` after the stage function finishes

Reply:
this method allows us to break out of the parent for loop thus we do not rerun a successful stage however you are perhaps right if we are looking for ease of code understanding